cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,79 @@ cuML is thread safe so its functions can be called from multiple host threads if
+1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - one can use [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) in C++11 to check POD types.
+2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`, instead.
+3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).
+Internally, for the C++ base layer atleast, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.
+
+### `libcuml++.so` (aka C++ API)
+Taking decisiontree-classifier algo as an example, the following way of exposing its API would be wrong according to the guidelines in this section. Because, this API exposes a non-POD C++ class object along the C++ API.
+```cpp
+template <typename T>
+class DecisionTreeClassifier {
+  TreeNode<T>* root;
+  DTParams params;
+  const cumlHandle &handle;
+public:
+  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
+  void fit(const T *input, int n_rows, int n_cols, const int *labels);
+  void predict(const T *input, int n_rows, int n_cols, int *predictions);
+};
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
+                                   DecisionTreeClassifier<float> *model, int n_rows,
+                                   int n_cols, int* predictions, bool verbose=false);
+```
+
+One of the right ways to expose this would be:
+```cpp
+// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
+// and passed between fit and predict methods
+template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
+struct DTParams { /* hyper-params for building DT */ };
+typedef TreeNode<float> TreeNodeF;
+typedef TreeNode<double> TreeNodeD;
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, TreeNodeF *&root, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
+                                   int n_cols, const TreeNodeD *root, int* predictions,
+                                   bool verbose=false);
+```
+The above example obviously under-plays the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.
+
+### WIP TODO (Need to explore this further!)
```

Comment:
I think it is worth considering to drop this section from the developer guide. IMHO it adds too much details making the guide too verbose.

Reply:
@vishalmehta1991 preferred to keep such examples.

Reply:
I think it is very useful to keep this section here, otherwise I can foresee receiving a PR where we need to request the store/load methods

Reply:
I am just concerned that a too verbose developer guide will not be read. In general I think we should try to keep it concise.

Reply:
I agree that a concise guide should be our aim. But I think we should keep this section for now and depending on how it gets received, we shall decide whether to keep it or not. What say?

Reply:
Fine with me.